GetEditorActivity Test

chinmay-kothari opened the pull request:
<!--
Before making a pull request, please:
1. Read the guidelines for contributing
1. Verify that your changes match our coding style
1. Fill out the requested information
-->

### Problem
<!-- What are you trying to solve? -->
Test case for GetEditorActivity function

### Solution
<!-- What does this PR do to fix the problem? -->
Test

### Areas of Impact
<!-- What parts of the codebase and which behaviors are affected? -->
Test dir

MonkeyDo commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,63 @@
+		const startDate = new Date();
+		startDate.setFullYear(2019, 1, 1);
+		const ActivityData = await getEditorActivity(editorJSON.id, startDate, Revision);
+		const MonthArray = ['Jan-19', 'Feb-19', 'Mar-19', 'Apr-19', 'May-19', 'Apr-19', 'May-19',
```
There's an extra `'Apr-19', 'May-19'` here, which points to an issue with the test: it doesn't test what's returned.

You can create an object with exactly the values you expect to get back from the utility, then use `expect(ActivityData).to.deep.equal(myExpectedResultsObject)`.
That way, you can test at the same time both if you have the right number of results *and* if you have one per month *and* if the month string and activityData are formatted correctly.

Also, by convention, variables and functions names use camelCase notation, while starting with an uppercase is used to represent class constructors, types and a few other special cases.

chinmay-kothari commented:
@MonkeyDo 
I have added another test as you mentioned for the condition when some months would be having multiple revisions while some would be with zero.

I made an `expectedResultObject` which mocks how we expect the result to be. Then compared it with the returned value. 

Let me know what you think about the changes :)

prabalsingh24 commented on test/src/server/routes/editor.js:
```diff
@@ -35,29 +36,83 @@ describe('getEditorActivity', () => {
+		// Adds remaining months upto the present month
+		const expectedStartDate = new Date();
+		expectedStartDate.setFullYear(2020, 0, 1);
+		const monthsUptoNow = eachMonthOfInterval({
```
Don't do like this. This will always give correct result because this is exactly the way it's been fetched in the `getEditorActivity` function.
Rather create an `expectedResult` object with hard coded values

chinmay-kothari replied:
The problem with creating hardcoded values is that our `getEditorActivity` function returns values till the present date.

So suppose if I do hardcode it till `Apr-20` then next month we'll need to add `May-20` so that the test does not fail next month. I personally don't think hard coding the values would be a good option. 

What else we can do here is pass an `endDate` into the `getEditorActivity` function default to the `Date.now()`. Then I think we can hardcode here with a precise end date. What do you think?

prabalsingh24 replied:
Oh yeah I didn't think of this. Nice catch!
Yeah I think passing an endDate would better.

MonkeyDo replied:
I agree, passing and endDate is the way to go, not only just for the purpose of testing but also for more granular use ("fetch my activity from january of this year")

prabalsingh24 commented on test/src/server/routes/editor.js:
```diff
@@ -35,29 +36,83 @@ describe('getEditorActivity', () => {
+		// Adds remaining months upto the present month
+		const expectedStartDate = new Date();
+		expectedStartDate.setFullYear(2020, 3, 1);
+		const monthsUptoNow = eachMonthOfInterval({
```
Same as above

prabalsingh24 commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,118 @@
+				new Revision(revisionAttribs).save(null, {method: 'insert'})
+			);
+		}
+		await Promise.all(RevisionArray);
```
RevisionArray => revisionsArray (Camel Case)

prabalsingh24 commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,118 @@
+describe('getEditorActivity', () => {
+	const {Revision} = orm;
+	let editorJSON;
+	const RevisionArray = [];
```
Instead of having a globally defined array, you can create a local array where it is used. Reusing the same array multiple times might cause some interferance

prabalsingh24 replied:
Also camel case.. and this is storing promises so maybe `revisionsPromiseArray` sounds better?

chinmay-kothari commented:
@prabalsingh24 
I have passed `endDate` into the `GetEditorActivity` with a default value of the present date. Also did the changes you mentioned. Let me know how it looks now :)

prabalsingh24 commented:
As @MonkeyDo mentioned, Test 1 and 2 can be merged.
Also how about having a test with an editor having 0 revisions? Just to make sure the function don't break in such cases

MonkeyDo commented on src/server/routes/editor.js:
```diff
@@ -207,7 +207,7 @@ async function getEditorActivity(editorId, startDate, Revision) {
 
 	const allMonthsInInterval = eachMonthOfInterval({
-		end: Date.now(),
+		end: endDate,
```
Addition of endDate is great.
I suggest one change that you can implement using TDD (test-driven development): first create a new test where you pass a string "bob" (or foo or bar or whatever :p) as the endDate.
I'm not sure what will happen with the current code, but you'll find out :)
Then, modify the test to expect it to throw a helpful error if the date is invalid.
Then, modify `getEditorActivity ` to act accordingly.
Finally, run the test again to ensure it's all working as expected.

It would be a good idea to do the same for startDate, now that I think of it.

You can use date-fns utilities to ensure the date is valid: https://date-fns.org/v2.12.0/docs/isValid#examples

chinmay-kothari commented:
> As @MonkeyDo mentioned, Test 1 and 2 can be merged.
> Also how about having a test with an editor having 0 revisions? Just to make sure the function don't break in such cases

I guess he said that for the previous two tests that were made, I have already merged those two. I also think having a test for zero revisions would be useful.

MonkeyDo commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,83 @@
+			authorId: editorJSON.id
+		};
+		const revisionsPromiseArray = [];
+		for (let i = 0; i < 12; i += 2) {
+			const tempDate = new Date();
+			tempDate.setFullYear(2020, i, 1);
+			revisionAttribs.id = random.number();
+			revisionAttribs.createdAt = tempDate;
+			revisionsPromiseArray.push(
+				new Revision(revisionAttribs).save(null, {method: 'insert'})
+			);
+		}
+		await Promise.all(revisionsPromiseArray);
```
If you modify fixtures created in `before` in a test, it's best to make sure it can't affect the tests that are run after it.
The easiest way to do that is to recreate the fixtures before each test by changing `before` and `after` functions to `beforeEach` and `afterEach`.

That way, for the next test you write, you don't have to keep track of the changes you might have made across the previous tests to know exactly what to test against. More deterministic is always better for tests.

chinmay-kothari commented:
@MonkeyDo,
For invalid dates, I have added error to be thrown and added their respective tests for checking the same. 
Let me know what you think of it. :)

chinmay-kothari commented:
> @MonkeyDo,
> For invalid dates, I have added errors to be thrown and added their respective tests for checking the same.
> Let me know what you think of it. :)

I guess it failed because of `random.number()` giving the same number..

MonkeyDo commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,142 @@
+	});
+
+	it('should throw an error for an invalid startDate', async () => {
+		const startDate = parseISO('bob');
```
parseIso will return specifically an Invalid Date object in this case.
Although it will not change the test or its results, I think startDate should be simply a string ("bob")

MonkeyDo commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,142 @@
+	it('should throw an error for an invalid endDate', async () => {
+		const startDate = new Date();
+		startDate.setFullYear(2020, 0, 1);
+		const endDate = parseISO('bob');
```
Same as above, let's keep it simple and pass a string

chinmay-kothari commented on test/src/server/routes/editor.js:
```diff
@@ -0,0 +1,142 @@
+		const activity = await getEditorActivity(editorJSON.id, startDate, Revision, endDate);
+
+		const expectedResult = {
+			// eslint-disable-next-line sort-keys
```
It would certainly clean up the code very well. Thanks

chinmay-kothari commented:
@MonkeyDo,
 I have done the changes you mentioned. Added `afterEach`, removed the comments, and passed the dates simply as a string. :p

MonkeyDo commented:
🎉 🚀
